General aarch64 cpu and hardware capabilities updates

bsdkurt opened the pull request:
* Simplify FreeBSD by using the values from `READ_SPECIALREG(midr_el1)` without use of the cpu tables.
* Switch to `elf_aux_info(3)` for hardware capabilities instead of reading special registers.
* Sync the HWCAP static asserts and auxv sanitation with Linux source
* Remove backward compat defines and extra structs and defines
* Update aarch64 cpu implementors and parts tables from FreeBSD & OpenBSD
* Switch strings in the cpu tables to use OpenBSD strings
* Use `READ_SPECIALREG(midr_el1)` on OpenBSD if `HWCAP_CPUID` is set, otherwise fallback to parsing `sysctl hw.model` using cpu tables.

This is a new version of the the diff. I noticed that FreeBSD didn't need to use the cpu tables and can just use the output from `midr_el1`. This should make FreeBSD future proof since it does not need the tables updated as new CPU's are supported. Also FreeBSD has supported reading from `READ_SPECIALREG(midr_el1)` for a long time but only recently set `HWCAP_CPUID` so that's why it doesn't check it on FreeBSD.

For OpenBSD, I can use `midr_el1` only when `HWCAP_CPUID` is set, otherwise for old CPUs, I fallback to parsing `sysctl hw.model` using the cpu tables. I'm hoping that all newer CPUs have `HWCAP_CPUID` and that I won't need to maintain the tables going forward.

I'm confident the FreeBSD changes are the right approach, but it needs a test build on aarch64. Whenever you have time - no rush.

bsdkurt committed:
General aarch64 cpu and hardware capabilities updates:
* Simplify FreeBSD by using the values from READ_SPECIALREG(midr_el1)
without use of the cpu tables.
* Switch to elf_aux_info(3) for hardware capabilities instead of
reading special registers.
* Sync the HWCAP static asserts and auxv sanitation with Linux source
* Remove backward compat defines and extra structs and defines
* Update aarch64 cpu implementors and parts tables from FreeBSD & OpenBSD
* Switch strings in the cpu tables to use OpenBSD strings
* Use READ_SPECIALREG(midr_el1) on OpenBSD if HWCAP_CPUID is set,
otherwise fallback to parsing sysctl hw.model using cpu tables.

bsdkurt committed:
Remove use of NULL that slipped in last commit.

bsdkurt commented:
There's a chance that the [commit](https://github.com/battleblow/jdk/commit/be8ad83f37110730fd2a68d99f6e64217851f286) for crashes on Apple silicon may not be needed after this lands. The old code only detected a subset of processor features on FreeBSD (NetBSD doesn't set any of them). The one I'm thinking might help is `HWCAP_ATOMICS`. See

https://github.com/battleblow/jdk/blob/bsd-port/src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp#L3128

snake66 commented:
I'm still working on getting the aarch64 reference machine set up correctly. Need to wait for some more packages to be installed. Hopefully I can get things to compile later today or tomorrow.

The patch looks clean, but need to go through it a bit more detailed.

bsdkurt committed:
* Provide implementations for get_current_sve_vector_length() and
set_and_get_current_sve_vector_length() using the minium supported
vector length of 128 bit / 16 bytes.
* Work-around certain CPUs that advertise SVE2 support but don't
actually suport SVE by only enabling SVE2 if SVE is also supported.
One such CPU is Apple's M4 processor.

bsdkurt commented:
I have a Macbook Pro M4 Pro with Macos on it. Using [UTM](https://getutm.app/) to test FreeBSD and OpenBSD on the M4 Pro processor that was reporting it supported SVE2.

This showed me that I needed to implement `VM_Version::get_current_sve_vector_length()` and  `VM_Version::set_and_get_current_sve_vector_length(int length)` for BSD. OpenBSD currently does not have an interface to query the sve vector length and FreeBSD appears to be working on it but not completed yet. There is a minimum sve vector length of 128 bits / 16 bytes. According to Mark Kettenis, an kernel developer who implement SVE on OpenBSD, most processors that do support SVE/SVE2 currently do so at the minimum 128 bit sve vector length. So the initial implementation for these functions is to just return the minimum. I have an XXX comment to remind us that FreeBSD can use the libsys interface when it is finished in the future.

Doing this test on an Apple M4 processor in a VM was somewhat fortunate because it revealed that both FreeBSD and OpenBSD currently report this processor supports SVE2. However, that is not strictly true and the jdk was getting SIGILL on both OpenBSD and FreeBSD trying to use SVE code on the M4. Quoting kettenis@ for a more complete explanation of the issue:

> So ARM messed up their architecture a bit with SVE2.  The SVEver field
> in the SVE feature ID register 0 (ID_AA64ZFR0_EL1) does not
> necessarily signal support for SVE2 when set to a non-zero value.
> 
> The architecture defines two modes: non-streaming SVE mode, which is
> what we usually call SVE and streaming SVE mode, which is also known
> as SME.  Now ARM clearly intended for SME to imply SVE support.
> However, Apple decided they wanted SME, but didn't bother to
> implement SVE.  So you have to check the SVE and SME fields in the
> ID_AA64PFR0_EL1 and ID_AA64PFR1_EL1 registers to determine what
> instructions are actually supported.
> 
> To get out if this mess, Linux actually changed the way HWCAP2_SVE
> (and other SVE-related bits in hwcap2) get set:
> 
>  https://lore.kernel.org/linux-arm-kernel/20250103142635.1759674-1-maz@kernel.org/  

He also suggested that the best work-around for this in the jdk would be to only allow SVE2 features to be used if the CPU also reports HWCAP_SVE is supported.

This now has been tested on the Apple M4 Pro in a qemu VM for both FreeBSD/14.3 and OpenBSD/7.8. I also tested the Apple M2 on OpenBSD/current (no SVE on M2). Mark will also do a test for me soon on Radxa Orion O6 that actually supports SVE (but not SVE2) on OpenBSD/current using a back-port of this diff to jdk25u. I'll report back his test results when I hear back.

snake66 commented:
@bsdkurt This is really nice work! I'm still waiting for the FreeBSD clusteradmins to install the packages I need on the reference server before I can test the aarch64 port there. (I have not pestered them over the holidays, but hope I'll get up and running soon now.)

I'll do a closer review and report back when I can test it.

bsdkurt committed:
Update comment to be more specific about the Apple M4 processor
and the included work-around.

bsdkurt commented:
@snake66 Sound good. Thanks for the update. I heard back from Mark on testing this on Radxa Orion O6 which supports SVE. All good there. He asked that I update the comment to be more specific about what the issue is on the M4 processor and SVE2.

Note that Linux kernel and now OpenBSD has changed how it reports HWCAP2_SVE2 and other SVE2 capabilities to only report them if the CPU also reports HWCAP_SVE:

https://lore.kernel.org/linux-arm-kernel/20250103142635.1759674-1-maz@kernel.org/
https://github.com/openbsd/src/commit/659bb8954e975876e3bc98812a0df46861933614

FreeBSD kernel devs may want to do the same so that it can better support the M4.

Edited to fix some typos where I put SME when I ment SVE.

bsdkurt commented:
> There's a chance that the [commit](https://github.com/battleblow/jdk/commit/be8ad83f37110730fd2a68d99f6e64217851f286) for crashes on Apple silicon may not be needed after this lands. The old code only detected a subset of processor features on FreeBSD (NetBSD doesn't set any of them). The one I'm thinking might help is `HWCAP_ATOMICS`. See
> 
> https://github.com/battleblow/jdk/blob/bsd-port/src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp#L3128

On the bsd-port branch (without this diff) I reverted that commit to see if I could reproduce the crashes on the Apple M4 running in a VM. I was not able to reproduce the crashes on the M4. Reading through the history of that commit, I see the problems were being reported on the Apple M1 processor. I only have M2 w/OpenBSD and M4 via qemu VM. There is no way for me to verify if this pull request will help the M1 processor that led to https://github.com/battleblow/jdk/commit/be8ad83f37110730fd2a68d99f6e64217851f286.

Oh well.

battleblow commented:
Sorry for the slow response.  Kurt, are you still looking for someone to give this a build on FreeBSD?  If so I can spin up an AWS instance and do so.

snake66 commented:
I have a Aarch64 instance via Hetzner up and running now, so I'll be able to test this PR quite soon. Just have to get some other stuff out of the way first.

bsdkurt commented:
> Sorry for the slow response. Kurt, are you still looking for someone to give this a build on FreeBSD? If so I can spin up an AWS instance and do so.

I was able to get one test in using FreeBSD 14.3 as a qemu VM guest on my Macbook with the M4 Max processor. The rest of the testing has been on OpenBSD with the following CPUs that I know of, Apple M4 Max (in VM), M2 Pro, ARM Cortex-A53, Radxa Orion O6 (w/SVE), Qualcomm Oryon r2p1. I have back ported this to OpenBSD's jdk ports 25-17 and so far no reports of fallout.

snake66 commented:
In addition to the expected SA test failures, I get a few failures in other parts that _could_ be interesting:

**Hotspot:**

* `compiler/c2/aarch64/TestStaticCallStub.java`: Calls to c2i interface stubs should be generated with near branches for segmented code cache up to 250MB
* `compiler/interpreter/VerifyStackWithUnreachableBlock.java`: Using VerifyStack for method that contains unreachable basic blocks compiler/loopopts/TestBackedgeLoadArrayFillMain.java: ArrayFill: if store is on backedge, last iteration is not to be executed.
* `compiler/loopopts/TestCountedLoopInsideInfiniteLoopMain.java`: Counted Loops inside infinite loops are only detected later, and may still be a Region and not a LoopNode as expected.
* `compiler/loopopts/TestDeadIrreducibleLoopsMain.java#id0`: Irreducible loops have many entries, only when the last entry loses control from the outside does the loop die, and have to disappear.
* `compiler/loopopts/TestDeadIrreducibleLoopsMain.java#id1`: Irreducible loops have many entries, only when the last entry loses control from the outside does the loop die, and have to disappear.
* `compiler/loopopts/TestDeadIrreducibleLoopsMain.java#id2`: Irreducible loops have many entries, only when the last entry loses control from the outside does the loop die, and have to disappear.
* `compiler/loopopts/TestDeadIrreducibleLoopsMain.java#id3`: Irreducible loops have many entries, only when the last entry loses control from the outside does the loop die, and have to disappear.
* `compiler/loopopts/TestInfiniteLoopWithUnmergedBackedgesMain.java`: Infinite loops may not have the backedges merged, before we call IdealLoopTree::check_safepts
* `compiler/loopopts/TestInlinedSplitFallInIrreducibleLoopStatusMain.java`: Regions that are inlined are by default tagged as NeverIrreducibleEntry.
* `compiler/loopopts/TestLongCountedLoopInInfiniteLoop.java`: C2: assert(!n->as_Loop()->is_loop_nest_inner_loop() || _loop_opts_cnt == 0) failed: should have been turned into a counted loop
* `compiler/loopopts/TestMoveStoreAfterInfiniteLoop.java`: C2: assert(!n_loop->is_member(get_loop(lca))) failed: control must not be back in the loop
* `compiler/loopopts/TestOnlyInfiniteLoopsMain.java`: Nested irreducible loops, where the inner loop floats out of the outer 
* `compiler/loopopts/TestPhaseCFGNeverBranchToGotoMain.java#id1`: Peeling of Irreducible loop can lead to NeverBranch being visited from either side
* `runtime/CompressedOops/CompressedClassPointers.java`: Testing address of compressed class pointer space as best as possible.
* `runtime/Locals/TestLargeLocalVarsStackRegion.java`:
* `runtime/interpreter/LastJsrTest.java`: Ensure a program that ends with a JSR does not crash

**Langtools:**

* `jdk/jshell/ErrorTranslationTest.java`: Tests for shell error translation
* `jdk/jshell/IdGeneratorTest.java`: Test custom id generators
* `jdk/jshell/KullaCompletenessStressTest.java`: Test SourceCodeAnalysis

I don't know if these failures are related to this patch or not, I'll give it another run without the patch for comparison.

This was on a machine with the following CPU info:

```
aarchie:~ $ dmesg|grep -iE '(cpu|arm)'
FreeBSD 15.0-RELEASE-p1 releng/15.0-n280999-7bceec30b351 GENERIC arm64
Starting CPU 1 (1)
Starting CPU 2 (2)
Starting CPU 3 (3)
Starting CPU 4 (4)
Starting CPU 5 (5)
Starting CPU 6 (6)
Starting CPU 7 (7)
FreeBSD/SMP: Multiprocessor System Detected: 8 CPUs
psci0: <ARM Power State Co-ordination Interface Driver> on acpi0
smccc0: <ARM SMCCC v1.1> on psci0
gic0: <ARM Generic Interrupt Controller v3.0> iomem 0x8000000-0x800ffff,0x80a0000-0x8ffffff on acpi0
its0: <ARM GIC Interrupt Translation Service> mem 0x8080000-0x809ffff on gic0
generic_timer0: <ARM Generic Timer> irq 34,35,36,37 on acpi0
Timecounter "ARM MPCore Timecounter" frequency 25000000 Hz quality 1000
Event timer "ARM MPCore Eventtimer" frequency 25000000 Hz quality 1000
cpu0: <ACPI CPU> on acpi0
armv8crypto0: <AES-CBC,AES-XTS,AES-GCM>
CPU  0: ARM Neoverse-N1 r3p1 affinity:  0
CPU  1: ARM Neoverse-N1 r3p1 affinity:  1
CPU  2: ARM Neoverse-N1 r3p1 affinity:  2
CPU  3: ARM Neoverse-N1 r3p1 affinity:  3
CPU  4: ARM Neoverse-N1 r3p1 affinity:  4
CPU  5: ARM Neoverse-N1 r3p1 affinity:  5
CPU  6: ARM Neoverse-N1 r3p1 affinity:  6
CPU  7: ARM Neoverse-N1 r3p1 affinity:  7
```

bsdkurt commented:
@snake66 I'm curious what hardware capabilities the Neoverse N1 supports. Can you run this test program that outputs the HWCAPS that the jdk cares about? The string matching on the hw.model wont match because I have OpenBSD strings in the tables now but the HWCAPS should print out ok on FreeBSD.

[jdk.aarch64.cpu.check.c](https://github.com/user-attachments/files/24531956/jdk.aarch64.cpu.check.c)

snake66 commented:
> @snake66 I'm curious what hardware capabilities the Neoverse N1 supports. Can you run this test program that outputs the HWCAPS that the jdk cares about?

Sure:

```
aarchie:~/src $ ./a.out
hw.model = ARM Neoverse-N1 r3p1
supports HWCAP_CPUID
cpu: 0x41, model 0xd0c, variant 0x3, rev: 0x1
lookup_cpu failed
cpu: 0x41, model 0x0, variant 0x3, rev: 0x1
HWCAP_FP
HWCAP_ASIMD
HWCAP_AES
HWCAP_PMULL
HWCAP_SHA1
HWCAP_SHA2
HWCAP_CRC32
HWCAP_ATOMICS
HWCAP_DCPOP
HWCAP_FPHP
HWCAP_ASIMDHP
```

Note that this is a VM running in a cloud, under QEMU, I think, so I'm not sure if that affects the report.

bsdkurt commented:
Here's the output from the M2 if you are curious:

```
hw.model = Apple Blizzard Pro r1p0
supports HWCAP_CPUID
cpu: 0x61, model 0x34, variant 0x1, rev: 0x0
HWCAP_FP
HWCAP_ASIMD
HWCAP_AES
HWCAP_PMULL
HWCAP_SHA1
HWCAP_SHA2
HWCAP_CRC32
HWCAP_ATOMICS
HWCAP_DCPOP
HWCAP_SHA3
HWCAP_SHA512
HWCAP_SB
HWCAP_PACA
HWCAP_FPHP
HWCAP_ASIMDHP
```

Has all the features of the Neoverse-N1 plus a few more.

snake66 commented:
> In addition to the expected SA test failures, I get a few failures in other parts that _could_ be interesting:

I get the same test failures on this machine when running the tests on the bsd-port branch, so they are at least not introduced by this patch. 

I don't have any other issues with this patch wither, so a thumbs up from me.
